third_party/move/move-compiler-v2/src/pipeline/split_critical_edges_processor.rs

```diff
@@ -0,0 +1,181 @@
+                self.emit_codes(codes)
+            },
+            (Some((l0_new, codes0)), Some((l1_new, codes1))) => {
+                self.emit(Bytecode::Branch(attr_id, l0_new, l1_new, t));
+                self.emit_codes(codes0);
+                self.emit_codes(codes1);
+            },
+        }
+    }
+
+    /// `label` is the target of some branch instruction
+    /// If label has multiple sources, returns
+    /// - a fresh label
+    /// - a new empty block that
+    ///     - starts with the fresh label
+    ///     - jumps to `label`
+    /// otherwise returns `None`
+    fn split_critical_edge(
+        &mut self,
+        attr_id: AttrId,
+        label: Label,
+    ) -> Option<(Label, Vec<Bytecode>)> {
+        // label is in `srcs_count` by construction
+        if *self.srcs_count.get(&label).expect("srcs count") > 1 {
+            Some(self.split_edge(attr_id, label))
+        } else {
+            None
+        }
+    }
+
+    /// Returns
+    /// - a fresh label
+    /// - a new empty block that
+    ///     - starts with the fresh label
+    ///     - jumps to `label`
+    fn split_edge(&mut self, attr_id: AttrId, label: Label) -> (Label, Vec<Bytecode>) {
+        let new_label = self.gen_fresh_label();
+        let codes = vec![
+            Bytecode::Label(attr_id, new_label),
+            Bytecode::Jump(attr_id, label),
+        ];
+        (new_label, codes)
+    }
+
+    /// Generates a fresh label
+    fn gen_fresh_label(&mut self) -> Label {
+        let new_label = Label::new(
+            if self.labels.is_empty() {
+                0
+            } else {
+                self.labels.iter().next_back().expect("label").as_usize() + 1
+            },
+        );
+        self.labels.insert(new_label);
+        new_label
+    }
+
+    fn emit(&mut self, bytecode: Bytecode) {
+        self.data.code.push(bytecode)
+    }
+
+    fn emit_codes(&mut self, codes: Vec<Bytecode>) {
+        for code in codes {
+            self.emit(code)
+        }
+    }
+}
+
+/// If key present in `map`, add 1 to its value; else insert 1
+fn map_inc<Key: Ord>(map: &mut BTreeMap<Key, usize>, key: Key) {
+    map.entry(key)
+        .and_modify(|n: &mut usize| {
+            let (n_suc, overflows) = n.overflowing_add(1);
+            if overflows {
+                panic!("`count_srcs` overflows")
+            } else {
+                *n = n_suc;
+            }
+        })
+        .or_insert(1usize);
+}
+
+/// Count the number of sources of labels
+/// labels with no sources are not included
+fn count_srcs(codes: &[Bytecode]) -> BTreeMap<Label, usize> {
+    let mut srcs_count = BTreeMap::new();
+    for code in codes {
+        match code {
+            Bytecode::Jump(_, label) => map_inc(&mut srcs_count, *label),
+            Bytecode::Branch(_, l0, l1, _) => {
+                map_inc(&mut srcs_count, *l0);
+                map_inc(&mut srcs_count, *l1);
+            },
```

Comment:
Can we have a fall-through flow to a label?  If not, please point me to docs describing why that can't be encountered here.

You also seem to be missing `AbortAction`.

Reply:
Is `AbortAction` a prover only thing? I checked `bytecode_generator` of compiler v2, and found all call instructions have no abort actions.

Reply:
Ok, but can't we still have a fall-through to a following block?  Shouldn't we handle that?

For example, see `label L4`  in the initial bytecode in https://github.com/aptos-labs/aptos-core/blob/main/third_party/move/move-compiler-v2/tests/unreachable-code-remover/break_unreachable.exp.

Reply:
done

Reply:
If you have to spend time digging through code ("checked the bytecode generator and ...") to discover an invariant, please be sure to document that invariant, both so others can save time in the future, and so that noone violates the accidental property you are relying upon.